### ISSUE TITLE
sendTicketGrantingTicketAction not removing primaryAuthN TGT

### ISSUE THREAD

**Issue by alexkolson** [Unicon/cas-mfa#128]
When logging in with Duo MFA, there are two TGT's generated in total. One after initial username/password form submission, and one after duo form submission.

Is this intended behavior?

If not, any pointers as to where I should look to start to debug/fix this issue?

**Comment by mmoayyed** [comment]
Make sure your cas servlet override xml file has the following:

```
<bean id="abstractAuthenticationViaFormAction" abstract="true"
          class="net.unicon.cas.mfa.web.flow.AbstractMultiFactorAuthenticationViaFormAction"
          p:centralAuthenticationService-ref="mfaAwareCentralAuthenticationService"
          p:multiFactorAuthenticationManager-ref="mfaAuthenticationManager"
          c:authenticationMethodRankingStrategy-ref="authenticationMethodRankingStrategy"
          c:authenticationMethodVerifier-ref="authenticationMethodVerifier"
          c:hostname="${server.prefix}"
          c:authenticationSupport-ref="authenticationSupport"
          c:multiFactorAuthenticationRequestResolver-ref="principalAttributeMfaRequestResolver" />

    <!--
      Initiating the the multifactor authentication sequence. Validates credentials and
      resolve principal, or authenticate and proceed to the second factor authentication flow.
    -->
    <bean id="initiatingAuthenticationViaFormAction" parent="abstractAuthenticationViaFormAction"
          class="net.unicon.cas.mfa.web.flow.InitiatingMultiFactorAuthenticationViaFormAction"
          c:wrapperAuthenticationAction-ref="authenticationViaFormAction"
          p:warnCookieGenerator-ref="warnCookieGenerator"/>

```

**Comment by mmoayyed** [comment]
As for pointers, you want to step through `InitiatingMultiFactorAuthenticationViaFormAction`.

**Comment by alexkolson** [comment]
When stepping through the flow, it looks as though both `initiatingAuthenticationViaFormAction` and `InitiatingMultiFactorAuthenticationViaFormAction` are both executing...or perhaps I am misinterpreting the log statements here: 

```
2015-12-08 09:08:14,748 DEBUG [org.springframework.webflow.execution.ActionExecutor] - Finished executing net.unicon.cas.mfa.web.flow.InitiatingMultiFactorAuthenticationViaFormAction@49297464; result = mfa-duo-two-factor
2015-12-08 09:08:14,748 DEBUG [org.springframework.webflow.execution.ActionExecutor] - Finished executing [EvaluateAction@2bae6fbb expression = initiatingAuthenticationViaFormAction, resultExpression = [null]]; result = mfa-duo-two-factor
```

Just trying to track it down and want to make sure I'm not way off base here.

**Comment by alexkolson** [comment]
Just more observations as I'm stepping though.

[`TerminatingMultiFactorAuthenticationViaFormAction`](https://github.com/Unicon/cas-mfa/blob/07c15a5ec26674ca471b5593924ec44adc367f47/cas-mfa-java/src/main/java/net/unicon/cas/mfa/web/flow/TerminatingMultiFactorAuthenticationViaFormAction.java) generates a TGT as well as one of the initiating form actions.

**Comment by alexkolson** [comment]
@mmoayyed Do we need to be calling submit on the `wrapperAuthentication` in `[`InitiatingMultiFactorAuthenticationViaFormAction` at line 91](https://github.com/Unicon/cas-mfa/blob/master/cas-mfa-java/src/main/java/net/unicon/cas/mfa/web/flow/InitiatingMultiFactorAuthenticationViaFormAction.java#L91)` since we authenticate the credentials down the flow in [`AbstractMultiFactorAuthenticationViaFormAction](https://github.com/Unicon/cas-mfa/blob/master/cas-mfa-java/src/main/java/net/unicon/cas/mfa/web/flow/AbstractMultiFactorAuthenticationViaFormAction.java#L183) at line 83`?

**Comment by mmoayyed** [comment]
Yes, you do. That's the call that executes/handles the first primary authn.

**Comment by mmoayyed** [comment]
Note that this is really a side-effect of CASImpl where authentication is tied to generation of TGT. We are handling this in 4.2 such that authentications are collected first, and then only a single TGT is generated at the end.

**Comment by alexkolson** [comment]
Ah that makes sense! @mmoayyed Following the flow all the way through I finally understand why this is happening. The only question I have now is it looks like the first ticket granting ticket is supposed to be destroyed by the overridden [`sendTicketGrantingTicketAction` at line 67](https://github.com/Unicon/cas-mfa/blob/master/cas-mfa-java/src/main/java/net/unicon/cas/mfa/web/flow/SendTicketGrantingTicketAction.java#L67), that is not happening for us and we have the destroyPreviousSSOSession flag set to true.

**Comment by mmoayyed** [comment]
There are a number of conditions built into that line. Are you able to step through and see which one fails?

**Comment by alexkolson** [comment]
Yeah it looks like the `ticketGrantingTicketValueFromCookie` is null inside of the `flowScope`. Logging these values shows:

```
2015-12-08 13:41:08,098 DEBUG [net.unicon.cas.mfa.web.flow.SendTicketGrantingTicketAction] - destroyPreviousSSOSession: true
,ticketGrantingTicketId (MFA TGT): TGT-2-M41byXtXpqLk19LivSd2NFpTc5DaPAoThNBQlOxlHesIp117SK-wyvern1
ticketGrantingTicketValueFromCookie (primary TGT): null
mfa credentials: net.unicon.cas.mfa.authentication.principal.MultiFactorCredentials@76d1c25c
```

This line:
`final String ticketGrantingTicketValueFromCookie = (String) context.getFlowScope().get("ticketGrantingTicketId");`

is returning null for some reason.

**Comment by mmoayyed** [comment]
Which is the code line that actually puts the TGT into the flowscope? That should happen as part of that wrapper object authenticating you. Note that that flowscope here belongs to the active webflow session. You have a primary session for the primary flow, and another session for the MFA subflow. Scan the context object in your debugger and see whether you can locate this attribute in any of the other flowscopes. There may be a mismatch. 

Are you able to duplicate this problem if you do a fresh clone off of master?

**Comment by alexkolson** [comment]
I'll attempt a fresh build off of master and see if the behavior still persists, although in the wrapper object authenticating me I never see a call to put the ticket in the flow scope only in the [request scope through `WebUtils`](https://github.com/Jasig/cas/blob/3.6.x/cas-server-core/src/main/java/org/jasig/cas/web/flow/AuthenticationViaFormAction.java#L109):

**Comment by alexkolson** [comment]
When I run a `git diff` from our master against unicon's master there are no differences and this still happens. Do you think it would be good enough for now to grab the TGT out of the request scope and throw it into the flowScope [around here in `InitiatingMultiFactorAuthenticationViaFormAction`](https://github.com/Unicon/cas-mfa/blob/master/cas-mfa-java/src/main/java/net/unicon/cas/mfa/web/flow/InitiatingMultiFactorAuthenticationViaFormAction.java#L96)?

**Comment by alexkolson** [comment]
Initial pass at a fix in #130.

**Comment by alexkolson** [comment]
@mmoayyed After adding that one line it appears to work just fine. What do you think about merging to master?

**Comment by mmoayyed** [comment]
I'll take a look tomorrow. Don't see any issues with merging this in though. I'll keep you posted. Thanks for diagnosing this.

**Comment by mmoayyed** [comment]
by the way you should know that I am in the process of converting the overlay to support 4.1.x The master branch will soon focus on 4.1, and 3.5 changes will move over to a different branch. I'll let you know when the switch happens, but just keep an eye out in case you were to submit future PRs. 

Thanks again.

**Comment by alexkolson** [comment]
Thanks for looking at it. I realize this issue isn't that big of deal, it'll just be nice not to have two sso sessions for every MFA authenticated principal.

**Comment by alexkolson** [comment]
@mmoayyed Roger that.

**Comment by alexkolson** [comment]
Closed in #130.
